source/fab/tools/shell.py

```diff
@@ -0,0 +1,46 @@
+##############################################################################
+# (c) Crown copyright Met Office. All rights reserved.
+# For further details please refer to the file COPYRIGHT
+# which you should have received as part of this distribution
+##############################################################################
+
+"""This file contains a base class for shells. This can be used to execute
+other scripts.
+"""
+
+from pathlib import Path
+from typing import List, Union
+
+from fab.tools.category import Category
+from fab.tools.tool import Tool
+
+
+class Shell(Tool):
```

Comment:
The general rule for scripting is to only ever use the Posix standard `sh`. That shell is always present on any Posix or Posix-like system and is always called and used the same way. This greatly improves interoperability.

As such I would recommend following suite and not trying to guess if a shell is present but to always use `sh`.

This may also mean the change to the `availability_option` type is unnecessary.

https://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html

Reply:
I clearly remember the UKMO having scripts with specific ksh-only syntax :)  So I would suggest that we try to support different shells, especially since it's basically no work. I will change the order in which shells are creates, so `sh` becomes the default.

The support for multiple parameters when checking availability was actually added for `sh`, since it does not support any version command :(

Reply:
That certainly happened and it was just as wrong then as it is now.

Clearly there is work to support multiple shells (and maintain that support) which we can avoid by not doing it.

As far as I can see fetching the version is only a means to determine if the shell is present. But `sh` is **always** present on Posix and Posix-like systems, so there is no need to check.

Reply:
> That certainly happened and it was just as wrong then as it is now.

No comment :)

I just accept that these specific scripts do exist.

> 
> Clearly there is work to support multiple shells (and maintain that support) which we can avoid by not doing it.
Not really, it's just a simple loop to create three different instances (and it is actually `sh` that caused all the additional coding overhead).


> 
> As far as I can see fetching the version is only a means to determine if the shell is present. But `sh` is **always** present on Posix and Posix-like systems, so there is no need to check.

I am not fetching a version number (I would if I could :) ), I am running 'echo hello' (or so). I don't know if sh is always present (I leave the posix-like out here ;) ). I just don't see what can hurt if we add two more shells in to cover more cases (and making it easy for any site to add `this-is-the-one-and-only-shell` to it ;)

Plus, I prefer all tools to be handled the same. Just assuming that `sh` always works seems wrong, any other tool also first checks if it can run some dummy command, which means we can get better feedback to the user.

Reply:
The reason for the link I sent was to show you the part of the Posix standard which *requires* the presence and capabilities of `sh`.

The reason supporting multiple shells hurts is because people might use different shells. What then happens if I try to run the build system on a platform which doesn't have that shell? It doesn't work is what happens.

This is why shell scripts should always be written using `sh`, the only standard shell. We can expect target platforms to have `sh` we can't expect them to have `bash`, `ksh` or `zsh`.

However I can see you have your heart set on this and frankly it isn't a hill worth dying on. We can always refactor it later if need be.

Reply:
> The reason supporting multiple shells hurts is because people might use different shells. What then happens if I try to run the build system on a platform which doesn't have that shell? It doesn't work is what happens.

This shell tool is not used at all by Fab, it is meant as a convenience for site-specific scripts. I've added this script since you didn't like the support of running compilers in a shell. This implies that  a site cannot use `$(nf-config -fflags)` to get for example the NetCDF flags, which we had assumed (other scripts besides nf-config are possible of course, including potentially using pkg-config this way - but that's another discussion

Therefore, site-specific scripts now need to run this command as a shell separately to detect for example the netcdf flags easily. So there is absolutely no risk that a script might not work for other sites, since other-sites won't use the same site-specific setup.

This is under the assumption that we will make sure that Fab does not add any bash/ksh/...-only scripts, which I intend to do.